Finer grained Dockerfile template

odoublewen opened the pull request:
This PR simply defines additional jinja "blocks" into various phases of the Dockerfile template.

If a user don't know about the blocks, doesn't need the blocks, doesn't care about the blocks.... then I am 99% sure this PR will have no effect whatsoever.  But having the blocks defined makes it much more convenient to extend the template.

This PR adds a total of 7 new blocks.  The only one I really need at the moment is the first one (`make_manifest`) but it seemed like a good and harmless thing to add the addiional blocks.

This is offered as a solution for the problem I described in https://github.com/spack/spack/issues/41656 and is an alternative solution to the one I proposed in https://github.com/spack/spack/pull/41558.

Commit:
Add more blocks to Dockerfile template

Commit:
Fix typo

Commit:
Fix line ordering

Commit:
Fix line ordering for proper jinja template nesting

odoublewen commented:
@wdconinc Sorry, one of the new blocks I introduced was mis-nested within an if-endif block.  The nesting should be correct now, and I expect the build tests to pass.

wdconinc commented:
Now that there are multiple blocks that can be overridden, could you extend the documentation at https://github.com/spack/spack/blob/develop/lib/spack/docs/containers.rst?plain=1#L516? (parsed into https://spack.readthedocs.io/en/latest/containers.html#extending-the-jinja2-templates)

I don't think it is necessary to provide an example for all blocks that can be extended, but we should just update the available blocks that can be overridden.

wdconinc commented:
We don't appear to have any unit tests for the extending of dockerfile templates (they would be in `lib/spack/spack/test`). Not required to add one.

Commit:
Add new dockerfile jinja blocks to docs

odoublewen commented:
@wdconinc I added some details to the docs regarding the new jinja blocks.  While I was doing so, I realized that the template would be even more convenient if it defined empty blocks at the end of the `build_stage` and `final_stage` blocks, so I took the liberty of adding `build_stage_extra_instructions` and `final_stage_extra_instructions`.  So these two (empty) stages basically replicate the functionality of `container.extra_instructions` [config](https://spack.readthedocs.io/en/v0.20.3/containers.html#creating-images-from-environments) that was removed by spack 0.21.0.

I realize the author of the template probably thought it was fine to do:
```
{% block final_stage %}
{{ super() }}
COPY data /share/myapp/data
{% endblock %}
```

...as opposed to:
```
{% block final_stage_extra_instructions %}
COPY data /share/myapp/data
{% endblock %}
```

And if that was the only difference, I would agree.  But if a user needs to override **multiple** inner blocks of the `build_stage`, then it becomes difficult to use the `{{ super() }}` approach.  The [extra example I added to the docs](https://github.com/spack/spack/pull/41746/files#diff-b56c695e4a61f1a56aed8a4c1daec756d28d77dfb38b37a5caed70f790133492R607-R621) shows what I mean.

Although the docs now show a bulleted list of all the extra "small block" names, I sort of think it would make sense to just link to the source of `share/spack/templates/container/Dockerfile` and say, "please refer to the Dockerfile template to see the available blocks".  Or maybe provide the list of blocks, and also provide the link.  I wasn't exactly sure how to link to a source file using sphinx (and get the link to come out right with github.)  What do you think?

Commit:
Update dockerfile template example to use super()

alalazo commented:
I'll double check, but if we added nested sub-blocks I kind of agree.

odoublewen commented:
Hi @alalazo Have you had a chance to consider the nested (optional) subblocks in this PR?

alalazo commented on lib/spack/docs/containers.rst:
```diff
@@ -591,6 +592,37 @@ the example below:
+.. code-block::
+   :emphasize-lines: 3-4,7
+
+   {% extends "container/Dockerfile" %}
+   {% block make_manifest %}
+   {{ super() }}
+   COPY custom_packages {{ paths.environment }}/custom_packages
+   {% endblock make_manifest %}
+   {% block final_stage_extra_instructions %}
+   COPY data /share/myapp/data
+   {% endblock %}
```
Isn't this case equivalent to:
```jinja
{% extends "container/Dockerfile" %}
{% block build_stage %}
COPY custom_packages {{ paths.environment }}/custom_packages
{{ super() }}
{% endblock build_stage %}
{% block final_stage %}
{{ super() }}
COPY data /share/myapp/data
{% endblock %}
```
?[^1] Asking to check whether the example shows something that needs inner blocks.

[^1]: in the sense that the copy operation can be moved before creating the `spack.yaml`.

odoublewen replied:
No, your suggestion does not work.  Because `mkdir {{ paths.environment }}` will throw an error if `{{ paths.environment }}` already exists.

See the issue that motivates this PR at https://github.com/spack/spack/issues/41656

alalazo replied:
Oh, I think that can be solved by using `mkdir -p` which is a much smaller fix

alalazo replied:
Ok, slowly putting the pieces together, I just saw #41558 :slightly_smiling_face:

odoublewen replied:
Yes, that was my first suggestion but wasn't liked by @wdconinc.  It's a "much smaller fix" in the sense of fewer characters commited to git, but I think @wdconinc's response was generally correct -- it opens up a potential for unintended consequences.

I think *this* fix is "much smaller" in the sense that it will have zero side-effect unless someone decides to use the blocks.

alalazo commented on lib/spack/docs/containers.rst:
```diff
@@ -591,6 +592,37 @@ the example below:
 
+The template also defines a number of inner blocks for finer-grained control over the rendered Dockerfile:
+
+* `make_manifest`
+* `install_software`
+* `strip_binaries`
+* `make_activate`
+* `build_stage_extra_instructions`
+* `copy_into_final`
+* `make_entrypoint`
+* `final_stage_extra_instructions`
+* `set_entrypoint`
```
If we add inner blocks, I would change the names slightly:
```suggestion
* `manifest`
* `install_software`
* `strip_binaries`
* `activation_script`
```
Also, the blocks:
- copy_into_final
- make_manifest
- set_manifest

are not inner blocks of "final_stage". I still have to think about names, but they should be documented above, in case.

alalazo replied:
Also, I would remove the empty `*_extra_instructions`, since the same can be obtained by extending the previous block.

odoublewen replied:
Yes, and I already suggested that, and it wasn't considered safe by @wdconinc.  https://github.com/spack/spack/pull/41558.  

All the details are in the first paragraph of this PR.

alalazo reviewed (changes requested):
In general, I'm not opposed to adding inner-blocks (as in, adding them isn't likely to cause any issue and it's backward compatible). 

That said, can you show an example where that is needed or clearly more convenient than extending the current blocks?

odoublewen commented:
Closing this as unnecessary since https://github.com/spack/spack/pull/41558 was merged
